hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java

```diff
@@ -1161,11 +1163,17 @@ private void stopAtException(Exception e){
     }
   }
 
-  protected HAState createHAState(StartupOption startOpt) {
+  protected HAState createHAState(Configuration conf) {
+    StartupOption startOpt = getStartupOption(conf);
     if (!haEnabled || startOpt == StartupOption.UPGRADE
         || startOpt == StartupOption.UPGRADEONLY) {
```

Comment:
Are there cases in which we'd want to start the observer with the `StartupOption.UPGRADE` or `StartupOption.UPGRADEONLY` options? When either of these are set this function returns `ACTIVE_STATE`.

Reply:
For both `StartupOption.Upgrade` and `StartupOption.UpgradeOnly` the documented behavior is to start a NameNode directly as Active, and then for UpgradeOnly it will shutdown after the upgrade steps are performed.

Reply:
Ok, then the code looks good to me.
Where can I find the documentation for the different options? The only comments I can find are these
https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java#L1088-L1090

Reply:
I'm not familiar with the upgrade code so I'm trying to figure it out as I review.

Reply:
Here is the document explaining the behavior and how it interacts with an HA cluster. https://hadoop.apache.org/docs/r3.3.4/hadoop-project-dist/hadoop-hdfs/HDFSHighAvailabilityWithQJM.html#HDFS_Upgrade.2FFinalization.2FRollback_with_HA_Enabled

Reply:
Thanks.